### PR TITLE
#28 improve health checks for bob

### PR DESCRIPTION
**Pull request by TimoKramer**
I am thinking about a better error message that clearly mentions the failing system, either docker or postgres. So people know what to do. what do you say?

### CODE REVIEW

File: src/bob/api/health_check.clj [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+;   Bob is distributed in the hope that it will be useful,
+;   but WITHOUT ANY WARRANTY; without even the implied warranty of
+;   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+;   GNU General Public License for more details.
+;
+;   You should have received a copy of the GNU General Public License
+;   along with Bob. If not, see <http://www.gnu.org/licenses/>.
+
+(ns bob.api.health-check
+  (:require [hugsql.core :as sql]
```
Comment (lispyclouds) [inline src/bob/api/health_check.clj]: this namespace is unused?

File: src/bob/api/health_check.clj [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  (docker/ping states/docker-conn))
+
+(defn pg-health-check
+  []
+  (db-health-check states/db))
+
+(defn health-check
+  "Check the systems we depend upon."
+  []
+  (d/let-flow [result (f/try-all [_ (docker-health-check)
```
Comment (lispyclouds) [inline src/bob/api/health_check.clj]: since these are one liner functions, we can inline them.

File: test/bob/api/middleware_test.clj [hunk 3]
```diff
@@ -13,7 +13,7 @@
 ;   You should have received a copy of the GNU General Public License
 ;   along with Bob. If not, see <http://www.gnu.org/licenses/>.
 
-(ns bob.middleware-test
+(ns bob.api.middleware-test
```
Comment (lispyclouds) [inline test/bob/api/middleware_test.clj]: is this change necessary?

**Comment by lispyclouds** [comment]
yes returning a custom message is quite helpful. you can do something like this in `try-all`:
```clojure
(when (f/failed? (f/try* (docker/ping states/conn)))
  (f/fail "Docker daemon is not healthy"))
```
similarly it can be used for pg. You can catch this message in the `(when-failed)` with `(f/message err)`. See https://github.com/bob-cd/bob/blob/master/src/bob/resource/core.clj#L74 as an example. Return this message with a 503

**Comment by lispyclouds** [comment]
Also would be nice to add debugging and error logs like https://github.com/bob-cd/bob/blob/master/src/bob/resource/internals.clj#L43

File: src/bob/api/health_check.clj (hunk 1 above)
Comment (TimoKramer) [reply src/bob/api/health_check.clj]: Are you asking, if sql is not used? Because line 25 uses sql...

Comment (lispyclouds) [reply src/bob/api/health_check.clj]: ah yes. missed that.

File: test/bob/api/middleware_test.clj (hunk 3 above)
Comment (TimoKramer) [reply test/bob/api/middleware_test.clj]: since I created another test-file in api-ns I moved middleware there, too. I wanted it to be symmetric like in src-folder.

Comment (lispyclouds) [reply test/bob/api/middleware_test.clj]: makes sense.

Comment (TimoKramer) [reply test/bob/api/middleware_test.clj]: src/bob/api/middleware.clj -> test/bob/api/middleware_test.clj
src/bob/api/health_check.clj -> test/bobt/api/health_check_test.clj

**Comment by TimoKramer** [comment]
when multiple systems are failing only the first one is communicated. Do you have an idea of a nice way to give out multiple failures?

**Comment by lispyclouds** [comment]
One way could be to define checks in a list and map over them. Filter this list for non failures are return this?

File: test/bob/api/health_check_test.clj [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+(ns bob.api.health-check-test
+  (:require [clojure.test :refer :all]
+            [failjure.core :as f]
+            [clj-docker-client.core :as docker]
+            [bob.api.health-check :refer :all]))
+
+(deftest health-check-various-conditions
+  (testing "all systems operational"
+    (with-redefs-fn {#'docker/ping (fn [x] "OK")
```
Comment (lispyclouds) [inline test/bob/api/health_check_test.clj]: constantly can be used here. See https://github.com/bob-cd/bob/blob/master/test/bob/pipeline/core_test.clj#L236

**Comment by lispyclouds** [comment]
if going for the list of checks way, you need to wrap each check with a
```clojure
(f/try* ... )
```

File: test/bob/api/health_check_test.clj (hunk 4 above)
Comment (TimoKramer) [reply test/bob/api/health_check_test.clj]: done

**Comment by lispyclouds** [comment]
this looks good, could you rebase and push too?

**Comment by TimoKramer** [comment]
I forgot to add logging, I will add that again.

**Comment by TimoKramer** [comment]
I think it is ok now... do you think the way I implemented the logging is fine? or would you prefer a let in the if-branch as well?

**Comment by lispyclouds** [comment]
this looks pretty good! Merging 🎉
